extensions/smallrye-context-propagation/runtime/src/main/java/io/quarkus/smallrye/context/runtime/SmallRyeContextPropagationProvider.java

```diff
@@ -0,0 +1,47 @@
+package io.quarkus.smallrye.context.runtime;
+
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Produces;
+import javax.inject.Singleton;
+
+import org.eclipse.microprofile.context.ManagedExecutor;
+import org.eclipse.microprofile.context.ThreadContext;
+
+import io.smallrye.context.impl.ManagedExecutorImpl;
+import io.smallrye.context.impl.ThreadContextImpl;
+
+@ApplicationScoped
+public class SmallRyeContextPropagationProvider {
+
+    private volatile ManagedExecutorImpl managedExecutor;
+
+    void initialize(ExecutorService executorService) {
+        managedExecutor = new ManagedExecutorImpl(-1, -1, (ThreadContextImpl) getAllThreadContext(), executorService, "no-ip") {
+            @Override
+            public void shutdown() {
+                throw new IllegalStateException("This executor is managed by the container and cannot be shut down.");
+            }
+
+            @Override
+            public List<Runnable> shutdownNow() {
```

Comment:
This behaviour is not expected by TCKs, see for instance https://github.com/eclipse/microprofile-context-propagation/blob/master/tck/src/main/java/org/eclipse/microprofile/context/tck/ManagedExecutorTest.java#L1737

Reply:
Here's the javadoc that says I must do this:

```java
 * <p>Managed executors that are created with the {@link Builder} or created for
 * injection into applications via CDI must support the life cycle methods, including:
 * awaitTermination, isShutdown, isTerminated, shutdown, shutdownNow.
 * The application must invoke <code>shutdown</code> or <code>shutdownNow</code> to terminate
 * the life cycle of each managed executor that it creates, once the managed executor is
 * no longer needed. When the application stops, the container invokes <code>shutdownNow</code>
 * for any remaining managed executors that the application did not already shut down.
 * The shutdown methods signal the managed executor that it does not need to remain active to
 * service subsequent requests and allow the container to properly clean up associated resources.</p>
 *
 * <p>Managed executors which have a life cycle that is scoped to the container,
 * including those obtained via mechanisms defined by EE Concurrency, must raise
 * IllegalStateException upon invocation of the aforementioned life cycle methods,
 * in order to preserve compatibility with that specification.</p>
```

Reply:
Defined on `ManagedExecutor`.

Reply:
That's not really what the TCK assumes though.
Also the bit you refer to is mostly for the sake of EE Conc compatibility which we don't really care about here. David's executor view allows to recognize that state of that ([here](https://github.com/jbossas/jboss-threads/blob/master/src/main/java/org/jboss/threads/ViewExecutor.java#L43-L46)), why not just forward the calls there?

Reply:
Well, then the TCK either does not test container-managed executors or it's invalid. That part of the spec is pretty important.

Now, I didn't know the view did not foward the shutdown to the underlying executor. In the case of container-managed executors that's fine (although it's contra spec which says it should throw), but I made it so SR-CP would use the view on any user-provided executor, and I'm not sure it's the right semantics here, because if the user creates a thread pool executor himself and we make a view out of it, the user may expect that shutting down our `ManagedExecutor` would also shut down the thread pool he created, but it does not. Do you think it's an issue @dmlloyd ?

Reply:
Hmm, that does sounds like an issue for those created through builders.
With beans, I suppose we could just weave in a disposer method that will perform this on bean destruction (which, with application scoped beans, is on application shutdown).is definitely

Or, we would have to make sure the threads in the pool are destroyed when not used (basically spawned on demand, dropped when idle for some time).

Reply:
> Or, we would have to make sure the threads in the pool are destroyed when not used (basically spawned on demand, dropped when idle for some time).

Uhm, hold on, that's actually how I did it back there. We should already have that in place here - https://github.com/smallrye/smallrye-context-propagation/blob/master/core/src/main/java/io/smallrye/context/SmallRyeManagedExecutor.java#L43-L49

Reply:
Filed https://github.com/eclipse/microprofile-context-propagation/issues/158 to clarify the spec/TCK mismatch, and possibly question the spec's decision about throwing.

Reply:
Well, the spec says:

    When the application stops, the container invokes <code>shutdownNow</code>
    for any remaining managed executors that the application did not already shut down.

So it looks like we need to keep hold of them all.

> Uhm, hold on, that's actually how I did it back there. We should already have that in place here

Yes but that's only for those that are created by the builder. Not for the views we create for user-specified executors (with the new SM-CP API to specify an executor in the builder).